Subject: [PATCH] Enable ohci1394 IRQ threading.
From: Sven-Thorsten Dietrich <sven@sven.thebigcorporation.com>
Date: Tue, 21 Oct 2008 02:01:21 -0700

Preliminary partitioning of the ohci1394 interrupt handler into
a quickcheck function that only disables the chip irq assert,
and a threaded main-handler function.

The status_cache allocation moves to the struct ti_ohci.

Signed-off-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
---
 drivers/ieee1394/ohci1394.c |   16 ++++++++++++++--
 drivers/ieee1394/ohci1394.h |    1 +
 2 files changed, 15 insertions(+), 2 deletions(-)

Index: linux-2.6.27-genirq-usb/drivers/ieee1394/ohci1394.c
===================================================================
--- linux-2.6.27-genirq-usb.orig/drivers/ieee1394/ohci1394.c
+++ linux-2.6.27-genirq-usb/drivers/ieee1394/ohci1394.c
@@ -2171,12 +2171,11 @@ static void ohci_schedule_iso_tasklets(s
 	spin_unlock_irqrestore(&ohci->iso_tasklet_list_lock, flags);
 }
 
-static irqreturn_t ohci_irq_handler(int irq, void *dev_id)
+static irqreturn_t ohci_irq_quickcheck(int irq, void *dev_id)
 {
 	quadlet_t event, node_id;
 	struct ti_ohci *ohci = (struct ti_ohci *)dev_id;
 	struct hpsb_host *host = ohci->host;
-	int phyid = -1, isroot = 0;
 	unsigned long flags;
 
 	/* Read and clear the interrupt event register.  Don't clear
@@ -2198,6 +2197,19 @@ static irqreturn_t ohci_irq_handler(int
 		return IRQ_NONE;
 	}
 
+	ohci->status_cache = event;
+	return IRQ_WAKE_THREAD;
+}
+
+static irqreturn_t ohci_irq_handler(int irq, void *dev_id)
+{
+	quadlet_t event, node_id;
+	struct ti_ohci *ohci = (struct ti_ohci *)dev_id;
+	struct hpsb_host *host = ohci->host;
+	int phyid = -1, isroot = 0;
+	unsigned long flags;
+
+	event = ohci->status_cache;
 	DBGMSG("IntEvent: %08x", event);
 
 	if (event & OHCI1394_unrecoverableError) {
Index: linux-2.6.27-genirq-usb/drivers/ieee1394/ohci1394.h
===================================================================
--- linux-2.6.27-genirq-usb.orig/drivers/ieee1394/ohci1394.h
+++ linux-2.6.27-genirq-usb/drivers/ieee1394/ohci1394.h
@@ -164,6 +164,7 @@ struct ti_ohci {
 
         /* remapped memory spaces */
         void __iomem *registers;
+	quadlet_t status_cache;  /* u32 saved by irq quickcheck */
 
 	/* dma buffer for self-id packets */
         quadlet_t *selfid_buf_cpu;
